include/igraph_datatype.h

```diff
@@ -30,6 +30,43 @@
 
 __BEGIN_DECLS
 
+struct igraph_i_property_cache_t;
+typedef struct igraph_i_property_cache_t igraph_i_property_cache_t;
+
+typedef enum {
+    /* Stores whether the graph has at least one self-loop. */
+    IGRAPH_PROP_HAS_LOOP = 0,
+
+    /* Stores whether the graph has at least one multi-edge, taking into account
+     * edge directions in directed graphs. In other words, this property should
+     * be false for a directed graph with edges (a, b) and (b, a), and true
+     * for a directed graph with edges (a, b) and (a, b) again. */
+    IGRAPH_PROP_HAS_MULTI,
+
+    /* Stores whether the graph has at least one reciprocal edge pair. Ignored
+     * in undirected graphs. This property should be true for a directed graph
+     * with edges (a, b) and (b, a), and false for a directed graph with
+     * edges (a, b) and (a, b) again. */
+    IGRAPH_PROP_HAS_RECIPROCAL,
+
+    /* Stores whether the graph is weakly connected. */
+    IGRAPH_PROP_IS_WEAKLY_CONNECTED,
+
+    /* Stores whether the graph is strongly connected */
+    IGRAPH_PROP_IS_STRONGLY_CONNECTED,
+
+    /* Stores whether the graph is a directed acyclic graph. Not used for
+     * undirected graphs. */
+    IGRAPH_PROP_IS_DAG,
+
+    /* Stores whether the graph is a forest, i.e. an undirected or directed
+     * graph that is cycle-free even if we ignore edge directions. */
+    IGRAPH_PROP_IS_FOREST,
+
+    /* Dummy value used to count enum values */
+    IGRAPH_PROP_I_SIZE
+} igraph_cached_property_t;
```

Comment:
Since we decided to leave the door open to adding non-bool properties in the future, should this type not have `bool` in the name as well? Right now the system works by assuming that these enum values can be used to index into a Boolean array.

I'm not sure what the best way is here, but if we want to design the API to allow this kind of extension in the future, it's best to think about it now.

Personally, I don't _think_ we'll ever need non-bool properties though (but as we've learned, these things are hard to predict).

Reply:
I think that the properties can be stored in a common enum, and if we start introducing properties with different types, we could have an array that maps property names into their storage types. Otherwise, if we started introducing different property enums for different cached property types, we'd need to have three variants of the `igraph_i_property_cache_has()` and the `igraph_i_property_cache_invalidate()` functions, one for each property enum.